File: Codes/Bingran Li/Sort/147. Insertion Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        // bug free
        if(head==null || head.next==null) return head;
        
        ListNode fakeHead = new ListNode(Integer.MIN_VALUE);
        fakeHead.next = head;
        
        ListNode sorted = head;
        ListNode current = head.next;
        
        while(sorted.next!=null){
            
            ListNode back = fakeHead;
            ListNode front = fakeHead.next;
            
            while(back!=sorted){
                
                if(current.val<front.val){
                    sorted.next = current.next;
                    back.next = current;
                    current.next = front;
                    current = sorted.next;
                    break;
                }
                back = front;
                front = front.next;
            }
            if(front == current){
                sorted = current;
                current = current.next;
            }
        }
        return fakeHead.next;
        
    }
}

File: Codes/Kan Zheng/C++/sort/147. Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        if(head==NULL)
            return NULL;
        ListNode* sudohead = new ListNode(-1);
        sudohead->next = head;
        ListNode* p_sec=sudohead;
        ListNode* p=head;
        ListNode* temp;
        while(p->next!=NULL)
        {
            while(p!=p_sec)
            {
                if((p->next->val<=p_sec->next->val))
                {
                    temp=p->next;
                    p->next = temp->next;
                    temp->next = p_sec->next;
                    p_sec->next=temp;
                    break;
                }
                else
                    p_sec=p_sec->next;
            }
            if(p==p_sec)
                p=p->next;
            p_sec = sudohead;
        }
        return sudohead->next;
    }
};